internal/integration/api/rotate.go

```diff
@@ -0,0 +1,158 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+//go:build integration_api
+
+package api
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/safe"
+
+	"github.com/siderolabs/talos/internal/integration/base"
+	"github.com/siderolabs/talos/pkg/machinery/api/common"
+	"github.com/siderolabs/talos/pkg/machinery/client"
+	"github.com/siderolabs/talos/pkg/machinery/config"
+	"github.com/siderolabs/talos/pkg/machinery/config/encoder"
+	"github.com/siderolabs/talos/pkg/machinery/config/generate/secrets"
+	"github.com/siderolabs/talos/pkg/machinery/config/machine"
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+	secretsres "github.com/siderolabs/talos/pkg/machinery/resources/secrets"
+	"github.com/siderolabs/talos/pkg/provision/access"
+	"github.com/siderolabs/talos/pkg/rotate/pki/talos"
+)
+
+// RotateCASuite verifies rotation of Talos and Kubernetes CAs.
+type RotateCASuite struct {
+	base.APISuite
+
+	ctx       context.Context //nolint:containedctx
+	ctxCancel context.CancelFunc
+}
+
+// SuiteName ...
+func (suite *RotateCASuite) SuiteName() string {
+	return "api.RotateCASuite"
+}
+
+// SetupTest ...
+func (suite *RotateCASuite) SetupTest() {
+	// make sure API calls have timeout
+	suite.ctx, suite.ctxCancel = context.WithTimeout(context.Background(), 5*time.Minute)
+}
+
+// TearDownTest ...
+func (suite *RotateCASuite) TearDownTest() {
+	if suite.ctxCancel != nil {
+		suite.ctxCancel()
+	}
+}
+
+// TestTalos updates Talos CA in the cluster.
+func (suite *RotateCASuite) TestTalos() {
+	if suite.Cluster == nil {
+		suite.T().Skip("cluster information is not available")
+	}
+
+	suite.T().Logf("capturing current Talos CA")
+
+	nodeInternalIP := suite.RandomDiscoveredNodeInternalIP(machine.TypeControlPlane)
+
+	// save osRoot
+	osRoot, err := safe.StateGetByID[*secretsres.OSRoot](client.WithNode(suite.ctx, nodeInternalIP), suite.Client.COSI, secretsres.OSRootID)
+	suite.Require().NoError(err)
+
+	suite.T().Logf("rotating current CA -> new CA")
+
+	newBundle, err := secrets.NewBundle(secrets.NewFixedClock(time.Now()), config.TalosVersionCurrent)
+	suite.Require().NoError(err)
+
+	options := talos.Options{
+		CurrentClient: suite.Client,
+		ClusterInfo:   access.NewAdapter(suite.Cluster),
+
+		ContextName: suite.Talosconfig.Context,
+		Endpoints:   suite.Client.GetEndpoints(),
+
+		NewTalosCA: newBundle.Certs.OS,
+
+		EncoderOption: encoder.WithComments(encoder.CommentsAll),
+
+		Printf: suite.T().Logf,
+	}
+
+	newTalosconfig, err := talos.Rotate(suite.ctx, options)
+	suite.Require().NoError(err)
+
+	newClient, err := client.New(suite.ctx, client.WithConfig(newTalosconfig))
+	suite.Require().NoError(err)
+
+	if !testing.Short() {
+		suite.restartAPIServices(newClient)
+	}
+
```

Comment:
shouldn't we assert cluster is healthy here also?

Reply:
it happens _inside_ the func, if !short, but the rotation process itself verifies connectivity to all apids.